packages/instant/src/components/scaling_input.tsx

```diff
@@ -0,0 +1,153 @@
+import * as _ from 'lodash';
+import * as React from 'react';
+
+import { ColorOption } from '../style/theme';
+import { util } from '../util/util';
+
+import { Input } from './ui';
+
+export enum ScalingInputPhase {
+    Start,
+    Scaling,
+    End,
+}
+
+export interface ScalingInputProps {
+    startWidthCh: number;
+    endWidthCh: number;
+    maxFontSizePx: number;
+    value?: string;
+    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
+    onFontSizeChange: (fontSizePx: number) => void;
+    fontColor?: ColorOption;
+    placeholder?: string;
+    maxLength?: number;
+}
+
+export interface ScalingInputState {
+    fixedWidthInPxIfExists?: number;
+}
+
+export interface ScalingInputSnapshot {
+    inputWidthPx: number;
+}
+// This is a magic number that was determined experimentally.
+const percentageToReduceByPerCharacter = 0.18;
+export class ScalingInput extends React.Component<ScalingInputProps, ScalingInputState> {
+    public static defaultProps = {
+        onChange: util.boundNoop,
+        onFontSizeChange: util.boundNoop,
+        maxLength: 9,
+    };
+    public state = {
+        fixedWidthInPxIfExists: undefined,
+    };
+    private readonly _inputRef = React.createRef();
+    public static getPhase(startWidthCh: number, endWidthCh: number, value?: string): ScalingInputPhase {
+        if (_.isUndefined(value) || value.length <= startWidthCh) {
```

Comment:
In my mind I thought of these phases as "None", "Scaling", and "Max", but I'm cool w/ Start, Scaling and End